operator/api/vectorized/v1alpha1/cluster_types.go

```diff
@@ -1418,3 +1408,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
+	out := make([]NodePoolSpec, 0)
+	if r.Spec.Replicas != nil {
+		defaultNodePool := NodePoolSpec{
+			Name:         nodepools.DefaultNodePoolName,
+			Replicas:     r.Spec.Replicas,
+			Tolerations:  r.Spec.Tolerations,
+			NodeSelector: r.Spec.NodeSelector,
+			Storage:      r.Spec.Storage,
+			Resources:    r.Spec.Resources,
+		}
+		if r.Spec.CloudStorage.CacheStorage != nil {
+			defaultNodePool.CloudCacheStorage = *r.Spec.CloudStorage.CacheStorage
+		}
+		out = append(out, defaultNodePool)
+	}
+	out = append(out, r.Spec.NodePools...)
+
+	return out
+}
+
+func (r *Cluster) CalculateCurrentReplicas() int32 {
+	var result int32
+	for _, np := range r.Status.NodePools {
+		result += np.CurrentReplicas
+	}
+	return result
+}
+
+func (r *Cluster) GetNodePools(ctx context.Context, k8sClient client.Reader) ([]*NodePoolSpecWithDeleted, error) {
+	var nodePoolsWithDeleted []*NodePoolSpecWithDeleted
+
+	nps := r.getNodePoolsFromSpec()
+	for i := range nps {
+		np := nps[i]
+
+		nodePoolsWithDeleted = append(nodePoolsWithDeleted, &NodePoolSpecWithDeleted{
+			NodePoolSpec: np,
+		})
+	}
+
+	// Also add "virtual NodePools" based on StatefulSets found.
+	// These represent deleted NodePools - they will not show up in spec.NodePools.
+	var stsList appsv1.StatefulSetList
+	err := k8sClient.List(context.TODO(), &stsList)
+	if err != nil {
+		return nil, err
+	}
+outer:
+	for i := range stsList.Items {
+		sts := stsList.Items[i]
+
+		if !strings.HasPrefix(sts.Name, r.Name) {
+			continue
+		}
+
+		for _, ownerRef := range sts.OwnerReferences {
+			if ownerRef.UID != r.UID {
+				continue outer
+			}
+		}
+
+		var npName string
+		if strings.EqualFold(r.Name, sts.Name) {
+			npName = nodepools.DefaultNodePoolName
+		} else {
+			npName = sts.Name[len(r.Name)+1:]
+		}
+
+		// Have seen it in NodePoolSpec
+		if slices.ContainsFunc(nps, func(np NodePoolSpec) bool {
+			return np.Name == npName
+		}) {
+			continue
+		}
+
+		replicas := sts.Spec.Replicas
+		if st, ok := r.Status.NodePools[npName]; ok {
+			replicas = &st.CurrentReplicas
+		}
+
+		var redpandaContainer *corev1.Container
+		for i := range sts.Spec.Template.Spec.Containers {
+			container := sts.Spec.Template.Spec.Containers[i]
+			if container.Name == "redpanda" {
+				redpandaContainer = &container
+				break
+			}
+		}
+		if redpandaContainer == nil {
+			return nil, fmt.Errorf("redpanda container not defined in STS %s template", sts.Name)
+		}
+
+		var datadirVcCapacity resource.Quantity
+		var datadirVcStorageClassName string
+
+		var cacheVcExists bool
+		var cacheVcCapacity resource.Quantity
+		var cacheVcStorageClassName string
+
+		for i := range sts.Spec.VolumeClaimTemplates {
+			vct := sts.Spec.VolumeClaimTemplates[i]
+			if vct.Name == "datadir" {
+				datadirVcCapacity = vct.Spec.Resources.Requests[corev1.ResourceStorage]
+				if vct.Spec.StorageClassName != nil {
+					datadirVcStorageClassName = *vct.Spec.StorageClassName
+				}
+			}
+			if vct.Name == "shadow-index-cache" {
```

Comment:
question: Does it makes sense to expose the const names from the statefulset pkg and use it here to avoid failure if these are renamed in the STS https://github.com/redpanda-data/redpanda-operator/blob/main/operator/pkg/resources/statefulset.go#L64-L65 ?

Reply:
good idea, done